src/stepper/stepper.ts

```diff
@@ -2455,9 +2479,168 @@ function treeifyMain(target: substituterNodes): substituterNodes {
   return treeify(target)
 }
 
+function jsTreeifyMain(target: substituterNodes, visited: Set<substituterNodes>): substituterNodes {
+  // recurse down the program like substitute
+  // if see a function at expression position,
+  //   visited before recursing to this target: replace with the name
+  //   else: replace with a CallExpression "(() => FunctionExpression)()" where the FunctionExpression's body is treeified
+  let verboseCount = 0
+  const treeifiers = {
+    // Identifier: return
+    ExpressionStatement: (target: es.ExpressionStatement): es.ExpressionStatement => {
+      return ast.expressionStatement(treeify(target.expression) as es.Expression)
```

Comment:
As much as possible, let's try not to use type assertions. They defeat the point of having a type system by silencing the compiler on type errors. Type assertions only affect the static semantics of a program but not the dynamic semantics, meaning that even though the TypeScript compiler doesn't complain, your program could run into runtime type errors. Instead, we should make use of [type narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html).

I know the current codebase isn't in a very good state (there are type assertions and non-null assertion operators everywhere), but let's try not to add onto the mess.

Reply:
This does require some effort. I open an issue for this so that it can hopefully be fully resolved. [issue #1457 ](https://github.com/source-academy/js-slang/issues/1457)